src/Microsoft.SqlTools.ServiceLayer/QueryExecution/DataStorage/SaveAsExcelFileStreamWriterHelper.cs

```diff
@@ -0,0 +1,583 @@
+            zipArchive.Dispose();
+        }
+
+
+        internal class ReferenceManager
+        {
+            public ReferenceManager(XmlWriter writer)
+            {
+                _writer = writer;
+            }
+            private int _currColumn; // 0 is invalid, the first AddRow will set to 1
+            private int _currRow = 1;
+            private char[] _currReference = new char[3 + 7]; //maximal XFD1048576
+            private int _currReferenceRowLength;
+            private int _currReferenceColumnLength;
+            private XmlWriter _writer;
+
+            public void AssureColumnReference()
+            {
+                if (_currColumn == 0)
+                {
+                    throw new ExporterException("AddRow must be called before AddCell");
+
+                }
+                if (_currColumn > 16384)
+                {
+                    throw new ExporterException("max column number is 16384, see https://support.office.com/en-us/article/Excel-specifications-and-limits-1672b34d-7043-467e-8e27-269d656771c3");
+                }
+            }
+
+            public void WriteAndIncreaseColumnReference()
+            {
+                _writer.WriteStartAttribute("r");
+                _writer.WriteChars(_currReference, 3 - _currReferenceColumnLength, _currReferenceRowLength + _currReferenceColumnLength);
+                _writer.WriteEndAttribute();
+                IncreaseColumnReference();
+            }
+
+            public void IncreaseColumnReference()
+            {
+                AssureColumnReference();
+                char[] reference = _currReference;
+                _currColumn++;
+                if ('Z' == reference[2]++)
+                {
+                    reference[2] = 'A';
+                    if (_currReferenceColumnLength < 2)
+                    {
+                        _currReferenceColumnLength = 2;
+                    }
+                    if ('Z' == reference[1]++)
+                    {
+                        reference[0]++;
+                        reference[1] = 'A';
+                        _currReferenceColumnLength = 3;
+                    }
+                }
+            }
+            private void ResetColumnReference()
+            {
+                _currColumn = 1;
+                _currReference[0] = _currReference[1] = (char)('A' - 1);
+                _currReference[2] = 'A';
+                _currReferenceColumnLength = 1;
+
+                string rowReference = _currRow.ToString();
+                _currReferenceRowLength = rowReference.Length;
+                rowReference.CopyTo(0, _currReference, 3, rowReference.Length);
+            }
+
+            public void AssureRowReference()
+            {
+                if (_currRow > 1048576)
+                {
+                    throw new ExporterException("max row number is 1048576, see https://support.office.com/en-us/article/Excel-specifications-and-limits-1672b34d-7043-467e-8e27-269d656771c3");
```

Comment:
We should put this string in SR.strings so that it can be localized. @benrr101 can you help with this after checkin? @wjzhou you can also do this but it does require Mono to be installed on your machine. In that case, add a new entry to SR.strings, save & run `.\build.cmd` which should add this to the SR.designer.cs file so you can reference it in your code.